File: lua/grains/grains.lua
local grains = {}

function grains.square(n)
  assert(n > 0 or n <= 64, "n must be between 1 and 64")
  return 2 ^ (n - 1)
end

function grains.total()
  local s = 0
  for i = 1,64 do
    s = s + grains.square(i)
  end
  return s
end

return grains

File: fsharp/grains/Grains.fs
module Grains
open System.Numerics

let square (n: int): Result<uint64,string> =
    match n with
    | n when n < 1 || n > 64 -> Error "square must be between 1 and 64"
    | _ -> Ok (uint64 (BigInteger.Pow(2, n-1)))

let total: Result<uint64,string> =
    let squares = [ for i in 1..64 -> square i]
    let sum = squares |> Seq.sumBy (fun x-> Result.defaultValue 0UL x)
    Ok (uint64 sum)
